Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+  }
+  /*! \brief Look-up with IndexAccumulator_Checked, const version.
+   */
+  helpers::IndexAccumulator_Checked<K, const NdFlattener<K,Data,Index> > checked() const {
+    return helpers::IndexAccumulator_Checked<K, const NdFlattener<K,Data,Index> >(size(),this,0);
+  }
+};
+
+template<typename _Data, typename _Index>
+class NdFlattener<1, _Data, _Index> {
+public:
+  static constexpr size_t K = 1;
+  using Data = _Data;
+  using Index = _Index;
+
+  typedef NdFlattener<1, Data, Index> CurrentLayer;
+  typedef CurrentLayer LowestLayer;
+
+private:
+  Index nNodes=0;
+  Index iNode=0;
+  std::vector<Data> data;
+
+
+  /*=== Getters ===*/
+public:
+  Data* GetData() {return data.data();}
+  const Data* GetData() const {return data.data();}
+
+  /*=== Outputting ===*/
+protected:
+  void toPythonString_fromto(std::ostream& output, Index from, Index to) const {
+    output  << "[";
+    for(Index i=from; i<to; ){
+      output << data[i];
+      if(++i<to) output << ", ";
+    }
+    output << "]";
+  }
+
+public:
+  NdFlattener(void) {}
+
+protected:
+  void allocate(){
+    data.reserve(nNodes);
+  }
+
+  void reset_iNode(){
+    iNode = 0;
+  }
+
+  /*=== Construct from 'recursive function' ===*/
+protected:
+  template<typename f_type>
+  void count_f(f_type f){
+    nNodes += f.first;
+  }
+
+  template<typename f_type>
+  void set_f(f_type f, bool refresh){
+    Index nChild = f.first;
+    for(Index iChild=0; iChild<nChild; iChild++){
+      data[iNode] = f.second(iChild);
+      iNode++;
+    }
+  }
+  
+  /*=== Construct with Allgatherv ===*/
+protected:
+  template<typename MPI_Environment_type>
+  void count_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version,
+         int const* displs, int const* ones)
+  { 
+    assert( Nodes_all[0]==nullptr);
+    Nodes_all[0] = new Index[mpi_env.size];
+    nNodes = 0;
+    // gather numbers of nodes in the current layer from all processes
+    mpi_env.MPI_Allgatherv( &(local_version->nNodes), 1, mpi_env.mpi_index, Nodes_all[0], ones, displs, mpi_env.mpi_index, mpi_env.comm );
+    for(int r=0; r<mpi_env.size; r++){
+      nNodes += Nodes_all[0][r];
+    }
+  }
+
+  template<typename MPI_Environment_type>
+  void set_g(MPI_Environment_type const& mpi_env,
+         Index** Nodes_all,
+         CurrentLayer const* local_version )
+  { 
+
+
+    int* Nodes_all_0_as_int = new int[mpi_env.size];
+    int* Nodes_all_0_cumulated = new int[mpi_env.size+1];
+    Nodes_all_0_cumulated[0] = 0;
+    for(int r=0; r<mpi_env.size; r++){
+      Nodes_all_0_cumulated[r+1] = Nodes_all_0_cumulated[r] + Nodes_all[0][r];
+      Nodes_all_0_as_int[r] = Nodes_all[0][r];
+    }
+
+    mpi_env.MPI_Allgatherv( local_version->data.data(), Nodes_all[0][mpi_env.rank], mpi_env.mpi_data, data.data(), Nodes_all_0_as_int, Nodes_all_0_cumulated, mpi_env.mpi_data, mpi_env.comm );
```

Comment:
please run clang_format on this file once you are done.

Reply:
done